### PR TITLE
Update iPhone iOS versions for SauceLabs

### PR DESCRIPTION
**Pull request by bencodeorg**
In the spirit of getting our automated test setup more current, move our automated iPhone tests from 11.3 to 15.4. This is one version behind the most recent available iOS version on SauceLabs, which is 16.0.

For a 15.x version, I'd rather do 15.6 (since 15.4 isn't commonly used per iPhone iOS stats below), but 15.6 isn't supported by Saucelabs.

Original motivation is that student learning is responsible for a long-time flaky UI test on iPhone (`can_see_finish`) that fails 26% of the time [at last check ](https://codedotorg.slack.com/archives/C03CM903Y/p1665083755135439). I ran this test on the old iOS version against my local server, and it succeeded 6/10 times. Against the newer iOS version, it succeeded 14/14 times. Doing a little probability, the probability of this happening by chance if the iOS version has nothing to do with it (given 26% flakiness) is about 1.5% (.74^14).

I considered a) updating to 16.0, and b) updating to iPad as well, but did not here for the following reasons:
- 16.0 involves a major version update to [Appium](https://saucelabs.com/platform/automation-tools/appium) from 1.x to 2.0, which broke some things -- eg, the device orientation config needs to be changed, names of tests showed up like `Unnamed job 700b7c11f0e843a2af9c97972e78780e`...felt out of scope for what motivated this change originally.
- 15.x on iPad has a bug (according to Saucelabs support) where a large menu covers half of the Safari browser window (see screenshot). They suggested using iOS 16 (eliminated for reason above) or <= 14. I've opened a [PR here to see if we could get iPad at least on 14.5](https://github.com/code-dot-org/code-dot-org/pull/48888) (most recent 14.x version on Saucelabs).

image

## Other considerations

All of this said, I am a little unclear on what versions of things we want SauceLabs to run. We just updated to Chrome 105, which is borderline current (106). We discussed our [Browser Version Support Strategy here](https://docs.google.com/document/d/17AIWp9Hzc2vwcVrH47aGU3kBLQbtiDCHfDvqrqEkE8A/edit), which has the following recommendation:

> **Update the browser versions we use for automated testing.** We should update the browser versions used by our automated tests to test the latest version (used by students and teachers) and the previous version representing P95 (the 95th percentile) of usage. Appendix A shows this for Google Chrome. Here, we would test against the latest version (v95) and the version at the 95th percentile (v87). Doing so ensures we are compatible with the latest version and many older versions that may still be in use by resource-constrained school districts.

Does this mean we ideally want to be running UI tests against two different versions of each browser?

## Links

- jira ticket: [SL-252](https://codedotorg.atlassian.net/browse/SL-252)
- [iPad OS stats (includes data from our non-prod servers, including test)](https://docs.google.com/spreadsheets/d/1D7wWQkEnvENEOs5VWIL9Ai38QMpq_trqEL68O1BHptU/edit#gid=804180480)
- [iPhone OS stats (includes data from our non-prod servers, including test)](https://docs.google.com/spreadsheets/d/1Aj1fGUzmGOxzvjIp37mv4azigjhKnPm1b0v1QVf9c90/edit#gid=461771574)

### CODE REVIEW

**Comment by cat5inthecradle** [comment]
Can you add a description to the PR? Maybe why these versions, why now?

**Review by pablo-code-org — APPROVED** [review]

**Comment by pablo-code-org** [comment]
talked with the infra team, and this is ok.

Is this because of the saucelabs issues in the previews days?

**Comment by pablo-code-org** [comment]
I approved without noticing @cat5inthecradle questions, let's follow up on those before merging :) We might have input on what version to use

**Comment by bencodeorg** [comment]
lol I was going to see if tests passed first and throw a proposal in #developers before doing formal review / description / etc. Sorry if github auto-requested review or something. I'll re-request review from y'all once I've done that!

**Comment by bencodeorg** [comment]
4 UI tests failed on the updated OS. Looking into it.

File: dashboard/test/ui/features/star_labs/can_see_finish.feature [hunk 1]
```diff
@@ -35,35 +35,30 @@ Feature: Make sure we can see the finish button for all LEVEL TYPE levels on sma
...
-  # The following levels currently do not show finish button
-  # in the viewport when play is pressed for small screens.
-
-  # TODO: Fix this - https://codedotorg.atlassian.net/browse/LP-1318
-    # "Minecraft Heroes Journey"
-    # "Minecraft Designer"
-    # "App Lab"
-
-  @no_ie @no_safari @no_firefox @no_chrome
+  # A row of arrow control buttons result in the Finish button not being visible in Dance Lab on our phone simulator,
```
Comment (bencodeorg) [inline dashboard/test/ui/features/star_labs/can_see_finish.feature:38]: This is a real issue on iPhone where you can't see the finish button on Dance Party. I'd rather get all of our iPhone tests on a more commonly used phone, then fix this bug (and re-enable this test).

File: dashboard/test/ui/features/star_labs/can_see_finish.feature [hunk 2]
```diff
@@ -35,35 +35,30 @@ Feature: Make sure we can see the finish button for all LEVEL TYPE levels on sma
...
+  # so only running against iPad.
+  # Tracking fix here: https://codedotorg.atlassian.net/browse/SL-346
+  @no_safari @no_firefox @no_chrome @no_phone
   Scenario: can see finish button on "Dance Party"
     And I check that the blockly free play level for "Dance Party" shows the finish button for mobile screens
 
-  @no_ie @no_safari @no_firefox @no_chrome
+  @only_mobile
```
Comment (bencodeorg) [inline dashboard/test/ui/features/star_labs/can_see_finish.feature:45]: This is shorthand for the previous test config (no real changes)

File: dashboard/test/ui/features/star_labs/contract_editor_sections.feature [hunk 3]
```diff
@@ -1,4 +1,4 @@
-@as_student
+@as_student @no_phone
```
Comment (bencodeorg) [inline dashboard/test/ui/features/star_labs/contract_editor_sections.feature:1]: This feature is used in /s/algebra, a very old lab which comprises 0.04% of our traffic (across all devices):

https://analytics.google.com/analytics/web/#/report/content-pages/a37745279w66217109p78966704/explorer-table.filter=algebra&explorer-table.plotKeys=%5B%5D

Didn't seem worth figuring out why this wasn't working on phone. FWIW it looked fine on my phone/Saucelabs.

File: dashboard/test/ui/features/star_labs/share_buttons.feature [hunk 4]
```diff
@@ -20,15 +20,15 @@ Feature: Share Buttons
 
   # Making DPad button show up in Game Lab and not in Sprite Lab for mobile
 
-  @no_ie @no_safari @no_firefox @no_chrome
+  @only_mobile
```
Comment (bencodeorg) [inline dashboard/test/ui/features/star_labs/share_buttons.feature:23]: Also just a shorthand update. This test was having some issues on iPad when I was initially trying to update both iPad and iPhone here (but no issues with just iPhone), but left the shorthand change for better readability.

**Comment by bencodeorg** [comment]
@pablo-code-org this is ready for review -- would love your take in particular on if there's any downside to having iPad and iPhone on different iOS versions.

@code-dot-org/studentlearning could use your input on if the tests I'm skipping seem ok

**Comment by breville** [comment]
Out of curiosity, what was the root cause of all those `can_see_finish` failures?  Was the button not visible in the failure cases because of the Safari sidebar showing?

**Comment by bencodeorg** [comment]
> Out of curiosity, what was the root cause of all those `can_see_finish` failures? Was the button not visible in the failure cases because of the Safari sidebar showing?

The Safari sidebar problem is only an issue on iPad 15.x. Flakiness was only observed on iPhone. I never root caused, but did observe consistent success with an upgrade to iOS 15.4 for iPhone tests.

**Review by molly-moen — APPROVED** [review]
LGTM!

**Pull request merged**